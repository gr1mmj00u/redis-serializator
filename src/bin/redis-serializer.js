#!/usr/bin/env node
import commander from 'commander';
import serializer from '..';

commander
  .version('1.0.2')
  .description('Redis serializer.')
  .command('keys <key> [otherKeys...]')
  .action((key, otherKeys = []) => {
    if (typeof key === 'undefined') {
      console.error('No key parameter.');
      process.exit(1);
    }

    const keys = [key, ...otherKeys];

    try {
      process.stdin
        .pipe(serializer(keys))
        .pipe(process.stdout);
    } catch (err) {
      console.error('Oops, something went wrong!');
      console.error(err.toString());
      process.exit(1);
    }
  });

commander.parse(process.argv);
