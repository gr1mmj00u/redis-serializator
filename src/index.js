import filter from './redis-serializer';

export default filter;
