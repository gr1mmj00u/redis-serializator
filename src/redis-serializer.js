import Parser from 'redis-parser';
import through2 from 'through2';

const excludedCmd = ['SET', 'SETNX'];

const replyArray = [];

const parser = new Parser({
  returnReply(reply) {
    replyArray.push(reply);
  },
  returnError(err) {
    throw err;
  },
  returnFatalError(err) {
    throw err;
  },
});

export const serialize = (arr = []) => {
  if (arr.length === 0) {
    return '';
  }

  const res = arr.reduce((acc, e) => [...acc, `$${e.length}`, e], []);

  return [`*${arr.length}`, ...res].join('\n');
};

const filter = keys => through2((data, enc, cb) => {
  parser.execute(data);

  const filteredReply = replyArray.filter(([cmd, k]) => !(excludedCmd.includes(cmd.toUpperCase())
        && keys.includes(k)));

  const result = filteredReply
    .reduce((acc, e) => [...acc, serialize(e)], [])
    .join('\n');

  cb(null, Buffer.from(result));
});

export default filter;
