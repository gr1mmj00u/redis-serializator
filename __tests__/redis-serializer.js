import fs from 'mz/fs';
import path from 'path';
import os from 'os';
import redisSerializer from '../src';
import { serialize } from '../src/redis-serializer';

const getFullPath = file => path.resolve(__dirname, file);

test('Serialize data', async () => {
  const res = serialize(['set', 'testKey', 'testValueKey']);
  expect(res).toMatchSnapshot();
});

test('Stream filter and serialize', async () => {
  const readStream = fs.createReadStream(getFullPath('__fixtures__/test-data.aof'), { encoding: 'utf-8' });

  const tempPath = await fs.mkdtemp(path.join(os.tmpdir(), 'foo-'));
  const tempFilePath = path.resolve(tempPath, 'test.aof');

  try {
    await readStream
      .pipe(redisSerializer(['test3', 'stas', 'stas22']))
      .pipe(fs.createWriteStream(tempFilePath, { encoding: 'utf-8' }));

    const data = await fs.readFile(tempFilePath, 'utf-8');

    expect(data).toMatchSnapshot();
  } catch (err) {
    console.log(err);
    expect(err).toMatch('error');
  }
});
