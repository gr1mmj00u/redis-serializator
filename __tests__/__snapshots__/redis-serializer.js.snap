// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Serialize data 1`] = `
"*3
$3
set
$7
testKey
$12
testValueKey"
`;

exports[`Stream filter and serialize 1`] = `
"*2
$6
SELECT
$1
0
*3
$3
SET
$5
stas2
$5
good2
*3
$3
SET
$4
test
$9
valueTest
*3
$3
SET
$5
test2
$10
valueTest2
*3
$3
SET
$5
stas4
$5
good4
*3
$3
SET
$5
stas3
$5
good3
*2
$6
SELECT
$1
0
*3
$3
set
$7
stas223
$8
stas2223
*3
$3
set
$7
stas223
$8
stas2223
*3
$3
set
$7
stas223
$8
stas2224
*3
$3
SET
$9
stasvasya
$9
staspetyz"
`;
